File: Backtracking/39.combination_sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

//
// Input: candidates = [2,3,6,7], target = 7,
//     A solution set is:
//     [
//         [7],
//         [2,2,3]
//     ]

// 很简单的backtrack的题目
var combinationSum = function(candidates, target) {
    if (candidates === null || candidates.length === 0) return [];
    candidates.sort(); // Make sure in ascending order

    let res = [];
    dfsHelper(res, candidates, target, [], 0, 0);
    return res;
};

const dfsHelper = function(res, candidates, target, curPath, sum, curIndex) {
    if (sum === target) {
        res.push(curPath.slice(0, curPath.length));
        return;
    } else {
        for (let i = curIndex; i < candidates.length; i++) {
            if (sum + candidates[i] <= target) {
                curPath.push(candidates[i]);
                dfsHelper(res, candidates, target, curPath, sum + candidates[i], i);
                curPath.pop();
            }
        }
    }
}


// 二刷
var combinationSum = function(candidates, target) {
    if (candidates === null) return [];
    let res = [];

    dfsHelper(res, candidates, target, [], 0, 0);
    return res;
};


const dfsHelper = function(res, candidates, target, curPath, curSum, curIdx) {
    if (curSum > target) return; //Since there is no negative number

    if (curSum === target) {
        res.push(curPath.slice(0, curPath.length));
    } else {
        let leftSum = target - curSum;
        for (let i = curIdx; i < candidates.length; i++) {
            if (candidates[i] < leftSum) {
                curPath.push(candidates[i]);
                dfsHelper(res, candidates, target, curPath, curSum + candidates[i], curIdx + 1);
                curPath.pop();
            }
        }
    }
}

File: Java/39.combination_sum.java
// Date: 11-13-2019
// Tag: Backtracking
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();

        if (candidates == null || candidates.length == 0) return res;
        Arrays.sort(candidates);
        List<Integer> path = new ArrayList<>();
        dfsHelper(res, path, candidates, 0, 0, target);
        return res;
    }

    public void dfsHelper(List<List<Integer>> res,
                          List<Integer> curPath,
                          int[] candidates,
                          int curIdx,
                          int curSum,
                          int target) {
        if (curSum > target) return;
        else if (curSum == target) {
            // System.out.println(curPath.toString());
            res.add(new ArrayList<>(curPath));
        } else {
            for (int i = curIdx; i < candidates.length; i++) {
                curPath.add(candidates[i]);
                dfsHelper(res, curPath, candidates, i, curSum + candidates[i], target);
                curPath.remove(curPath.size() - 1);
            }
        }
    }
}
